src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/CopyDataMessage.java

```diff
@@ -29,19 +32,32 @@
   protected static final char IDENTIFIER = 'd';
 
   private byte[] payload;
+  private CopyStatement statement;
 
   public CopyDataMessage(ConnectionHandler connection) throws Exception {
     super(connection);
+    // Payload byte array excluding 4 bytes containing the length of message itself
     this.payload = new byte[this.length - 4];
     if (this.inputStream.read(this.payload) < 0) {
       throw new IOException("Could not read copy data.");
     }
+    this.statement = (CopyStatement) connection.getActiveStatement();
   }
 
   @Override
   protected void sendPayload() throws Exception {
-    throw new IllegalStateException(
-        "Spanner does not currently support the copy functionality through the proxy.");
+    // If backend error occurred during copy-in mode, drop any subsequent CopyData messages.
+    MutationWriter mw = this.statement.getMutationWriter();
+    if (!statement.hasException()) {
+      try {
+        mw.buildMutation(this.connection, this.payload);
+      } catch (SQLException e) {
+        statement.handleExecutionException(e);
```

Comment:
If we don't write to the error file here, couldn't we potentially lose rows? Although, I guess it's risky to write because we could also end up with  duplicate rows depending on whether the error occurred before or after the rows were sent to Spanner

Reply:
We could write to the error file here on an exception if we wait until CopyDone to actually write to spanner.  Everything should be rolled back on a failure/error, so I think it is best to wait for CopyDone to do the write/commit.

Reply:
agreed. The current implementation that has sendPayload inside of CopyDataMessage is based on the stretch  goal that we plan to revert. I think this looks good for now. In the next CL let's make sure to get all of the non-stretch goal logic correct.

1. we will no longer track the 64KB batch size. Instead, we'll track the 100MB mutation API commit size limit and continue tracking the 20K mutation count limit. If either is exceeded, CopyDataMessage should write everything to the error file and return an error
2. In CopyDone, if none of the limits have been exceeded, we'll write to Spanner. Note that this can still fail because the mutation count inside of PGAdapter is an approximation and is not accurate if the table has indexes or is interleaved with other tables. If the write to Spanner fails, CopyDone should write everything to the error file and return an error